src/AggregateFunctions/QuantileTDigest.h

```diff
@@ -213,15 +273,23 @@ class TDigest
         size_t size = 0;
         readVarUInt(size, buf);
 
-        if (size > params.max_unmerged)
+        if (size > max_centroids_deserialize)
             throw Exception("Too large t-digest centroids size", ErrorCodes::TOO_LARGE_ARRAY_SIZE);
 
+        count = 0;
+        unmerged = 0;
+
         centroids.resize(size);
+        // From now, TDigest will be in invalid state if exception is thrown.
         buf.read(reinterpret_cast<char *>(centroids.data()), size * sizeof(centroids[0]));
 
-        count = 0;
         for (const auto & c : centroids)
+        {
+            if (c.count <= 0 || std::isnan(c.count) || std::isnan(c.mean)) // invalid count breaks compress(), invalid mean breaks sort()
+                throw std::runtime_error("Invalid centroid " + std::to_string(c.count) + ":" + std::to_string(c.mean));
```

Comment:
It'd better to throw `Exception(<error_message>, ErrorCodes::<some_error>);`. Choose an appropriate error code from the list (https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/ErrorCodes.cpp) and add it in `namespace ErrorCodes` at the beginning of this file like `TOO_LARGE_ARRAY_SIZE`.